### PR TITLE
Update image info schema to have image/platform concepts

### LINKED ISSUE #286: Image info format doesn't separate metadata for multiples images built from the same Dockerfile
The [image info](https://github.com/dotnet/versions/blob/eb5a38e95113c07585dab308669f084e492d16dd/build-info/docker/image-info.dotnet-dotnet-docker-master.json#L737) file keys off of the image's Dockerfile folder path.  In the case of the sample projects, there can be a single Dockerfile that is built on multiple platforms, each producing different tags.  So when the image info file gets updated with the metadata from these images, that metadata all goes into the same element since the images all share the same Dockerfile folder path.  This doesn't currently cause any issues with the image dependency flow system because it only looks up the digests for the images that it cares about and ignores the extraneous other data in the image info file.  But from a correctness standpoint, the metadata contained in the image info file should be separated properly.

### PR DESCRIPTION
**Pull request by mthalman**
In order to support data like shared tags in the image info file, the schema needs to be updated to group Dockerfiles associated with a single manifest tag just as the manifest does.  This grouping is referred to as an image while each of the underlying items are referred to as a platform.  I've updated the schema to reflect this by renaming the previously named `ImageData` to `PlatformData` because that is really what it represents now and introducing a new `ImageData` class that groups the `PlatformData` instances.

OLD schema example:
```
[
  {
    "repo": "repo2",
    "images": [
      {
        "1.0/repo2/os1/Dockerfile": {
          "simpleTags": [
            "tag1"
          ],
        },
        "1.0/repo2/os2/Dockerfile": {
          "simpleTags": [
            "tag2"
          ],
        }
      }
    ]
  }
]
```

NEW schema example:
```
[
  {
    "repo": "repo2",
    "images": [
      {
        "platforms": {
          "1.0/repo2/os1/Dockerfile": {
            "simpleTags": [
              "tag1"
            ]
          }
        }
      },
      {
        "platforms": {
          "1.0/repo2/os2/Dockerfile": {
            "simpleTags": [
              "tag2"
            ]
          }
        }
      }
    ]
  }
]
```

The main impact this change causes is with the merging behavior.  There needs to be a way to determine which image instance a platform belongs when merging a source to a target.  The image has no identifying data.  This is accomplished by correlating in memory each image in an image info file with its related image in the manifest file.  That can then be used as a reference identifier to pair up related images between two separate image info files.

Fixes #286

### CODE REVIEW

**Comment by MichaelSimons** [comment]
The proposed schema makes me think there is an existing issue with the current schema that is not addressed with the new schema.  That is using the Dockerfile path as the key to the image/platform Dictionary is an issue.  Specially in the case of the samples where we build the same Dockerfile multiple time - https://github.com/dotnet/versions/blob/master/build-info/docker/image-info.dotnet-dotnet-docker-master.json#L1477.  To address this, should the image collection be an array and the Dockerfile be an attribute of the platform?

**Comment by MichaelSimons** [comment]
https://github.com/dotnet/docker-tools/issues/286 tracks the Dockerfile path issue I previously mentioned.  I think we should be addressing this problem with this schema change.

**Comment by mthalman** [comment]
> The proposed schema makes me think there is an existing issue with the current schema that is not addressed with the new schema. That is using the Dockerfile path as the key to the image/platform Dictionary is an issue. Specially in the case of the samples where we build the same Dockerfile multiple time - https://github.com/dotnet/versions/blob/master/build-info/docker/image-info.dotnet-dotnet-docker-master.json#L1477. To address this, should the image collection be an array and the Dockerfile be an attribute of the platform?

Yes, I've made that change.  This required adding some more data to the `PlatformData` class to allow a given instance to be distinguishable from another that has the same Dockerfile path.  So the distinguishing data of a platform is Dockerfile path, architecture, OS type, and OS version.

File: src/Microsoft.DotNet.ImageBuilder/src/Models/Image/ImageData.cs [hunk 1]
```diff
@@ -2,25 +2,40 @@
...
+
+            if (ManifestImage == other.ManifestImage)
+            {
+                return 0;
+            }
 
-        public string Digest { get; set; }
+            // If we're comparing two different image items, compare them by the first Platform path's to
+            // provide deterministic ordering.
+            return Platforms.FirstOrDefault()?.Path.CompareTo(other.Platforms.FirstOrDefault()?.Path) ?? 1;
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Microsoft.DotNet.ImageBuilder/src/Models/Image/ImageData.cs]: Updated to use Platform instance from comparison.

File: src/Microsoft.DotNet.ImageBuilder/src/ImageInfoHelper.cs [hunk 2]
```diff
@@ -5,15 +5,60 @@
...
+                {
+                    // A given Dockerfile path is unique to an image. Take one of those Dockerfile paths
+                    // from the platform of the image info model and find the manifest image that contains
+                    // that same Dockerfile path.
+                    string dockerfilePath = imageData.Platforms.FirstOrDefault()?.Path;
+                    if (dockerfilePath != null)
+                    {
+                        foreach (ImageInfo manifestImage in manifestRepo.AllImages)
+                        {
+                            // This assumes that the manifest will never have two platforms with the same Dockerfile path contained
```
[the remark this answers is not in the document]
Comment (mthalman) [reply src/Microsoft.DotNet.ImageBuilder/src/ImageInfoHelper.cs]: Good point.  Updated this to compare by the full state that distinguishes a platform.

**Comment by mthalman** [comment]
> LGTM - Any thoughts on my previous suggestion to use a feature branch?

Oh right, thanks for the reminder.  I'll do that.

**Pull request merged**